File: learn-julia/old-learn-julia/comparison/matmul.py
from random import random
import time


def printvec(vec):
    print("[", end="")
    for i in range(len(vec) - 1):
        print(f"{vec[i]:.6f}, ", end="")
    print(f"{vec[len(vec)-1]:.6f}", end="")
    print("]")


def printmat(mat):
    for i in range(len(mat)):
        printvec(mat[i])


def randmat(m, n):
    return [[random() for _ in range(n)] for _ in range(m)]


def zeromat(m, n):
    return [[0 for _ in range(n)] for _ in range(m)]


def matmul(A, B):
    n_rows_A = len(A)
    n_cols_A = len(A[0])

    n_rows_B = len(B)
    n_cols_B = len(B[0])

    if n_cols_A != n_rows_B:
        raise RuntimeError("Cannot multiply mismatched matrices!")

    n_rows_C = n_rows_A
    n_cols_C = n_cols_B
    C = zeromat(n_rows_C, n_cols_C)
    for i in range(n_rows_C):
        for j in range(n_cols_C):
            C[i][j] = 0
            for k in range(n_cols_A):
                C[i][j] += A[i][k] * B[k][j]
    return C


def main():
    m, n, p = 500, 500, 500
    with open("py_myimpl.txt", "wt") as f:
        for _ in range(100):
            A = randmat(m, n)
            B = randmat(n, p)

            start = time.process_time()
            C = matmul(A, B)  # noqa
            end = time.process_time()

            proc_time = end - start
            print(proc_time, file=f)

    print("Done")


if __name__ == "__main__":
    main()

File: learn-julia/old-learn-julia/comparison/matmul.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct Matrix {
  float **X;
  int numRows;
  int numCols;
} Matrix;

void printvec(float *x, int n) {
  printf("[");
  for (int i = 0; i < n - 1; i++) {
    printf("%f, ", x[i]);
  }
  // last element in the row should not have a trailing comma
  printf("%f", x[n - 1]);
  printf("]\n");
}

void printmat(Matrix mat) {
  for (int i = 0; i < mat.numRows; i++) {
    printvec(mat.X[i], mat.numCols);
  }
}

Matrix randmat(int m, int n) {
  Matrix mat;
  mat.X = (float **)malloc(sizeof(float *) * m);
  for (int i = 0; i < m; i++) {
    mat.X[i] = (float *)malloc(sizeof(float) * n);
    for (int j = 0; j < n; j++) {
      mat.X[i][j] = (float)random() / RAND_MAX;
    }
  }
  mat.numRows = m;
  mat.numCols = n;
  return mat;
}

Matrix zeromat(int m, int n) {
  Matrix mat;
  mat.X = (float **)malloc(sizeof(float *) * m);
  for (int i = 0; i < m; i++) {
    mat.X[i] = (float *)malloc(sizeof(float) * n);
    for (int j = 0; j < n; j++) {
      mat.X[i][j] = 0.0f;
    }
  }
  mat.numRows = m;
  mat.numCols = n;
  return mat;
}

Matrix matmul(Matrix A, Matrix B) {
  if (A.numCols != B.numRows) {
    printf("Unable to multiply mismatched matrices!");
    exit(-1);
  }

  Matrix C = zeromat(A.numRows, B.numCols);
  for (int i = 0; i < C.numRows; i++) {
    for (int j = 0; j < C.numCols; j++) {
      C.X[i][j] = 0;
      for (int k = 0; k < A.numCols; k++) {
        C.X[i][j] += A.X[i][k] * B.X[k][j];
      }
    }
  }
  return C;
}

void freemat(Matrix M) {
  for (int i = 0; i < M.numRows; i++) {
    free(M.X[i]);
  }
  free(M.X);
}

int main(int argc, char **argv) {
  int m = 500;
  int n = 500;
  int p = 500;

  Matrix A, B, C;
  struct timespec start, end;

  FILE *f = fopen("c_myimpl.txt", "wt");
  if (f == NULL) {
    printf("Unable to open file!");
    exit(-1);
  }

  for (int i = 0; i < 100; i++) {
    A = randmat(m, n);
    B = randmat(n, p);

    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &start);
    C = matmul(A, B);
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &end);

    long nanosecs =
        (end.tv_sec * 1e9 + end.tv_nsec) - (start.tv_sec * 1e9 + start.tv_nsec);
    double secs = nanosecs / 1e9;
    fprintf(f, "%lf\n", secs);

    freemat(A);
    freemat(B);
    freemat(C);
  }

  fclose(f);
  printf("\nDone.\n");
}
